File: kotlin-automation/src/test/kotlin/com/myecommerce/automation/driver/api/MyEcommerceDriver.kt
package com.myecommerce.automation.driver.api

import com.myecommerce.automation.driver.ports.HttpPort
import com.myecommerce.automation.driver.ports.getAs
import com.myecommerce.automation.driver.ports.getListAs
import com.myecommerce.automation.driver.ports.getListWithQueryAs
import com.myecommerce.automation.dsl.domain.DeliveryOption
import com.myecommerce.automation.dsl.domain.DeliveryState
import com.myecommerce.automation.dsl.domain.ProductCard
import com.myecommerce.automation.dsl.domain.ProductDetail
import com.myecommerce.automation.dsl.domain.ProductListing
import com.myecommerce.automation.dsl.domain.SearchResults
import com.myecommerce.automation.dsl.protocols.CatalogueProtocol
import com.myecommerce.automation.dsl.protocols.Channel
import com.myecommerce.automation.dsl.protocols.DriverRegistry

class MyEcommerceDriver(private val http: HttpPort) : CatalogueProtocol {

    private var currentProductId: Int = -1
    private var lastSearchTerm: String = ""

    companion object {
        private const val BASE_URL = "http://localhost:8001"

        init {
            DriverRegistry.register(Channel.API) {
                val http: HttpPort = if ("okhttp" == System.getProperty("http.impl", "restassured"))
                    OkHttpHttpPort(BASE_URL)
                else
                    RestAssuredHttpPort(BASE_URL)
                MyEcommerceDriver(http)
            }
        }
    }

    override fun browseCatalogue() {}
    override fun returnToProductListing() {}
    override fun chooseAlternativeDeliveryOption() {}

    override fun viewFirstProduct() {
        currentProductId = fetchProducts("").first().id
    }

    override fun searchFor(term: String) {
        lastSearchTerm = term
    }

    override fun getProductListing(): ProductListing {
        val cards = fetchProducts("").map { it.toProductCard() }
        return ProductListing(cards, hasVisibleLoadingIndicators = false)
    }

    override fun getProductDetail(): ProductDetail {
        val product = fetchProductDetail(currentProductId)
        return ProductDetail(
            title = product.title,
            price = formatPrice(product.price),
            description = product.description,
            imagePresent = !product.imageUrl.isNullOrBlank(),
            addToCartButtonText = "Add to Cart",
            addToCartEnabled = true,
        )
    }

    override fun getDeliveryState(): DeliveryState {
        val activeOptions = fetchActiveDeliveryOptions()
        if (activeOptions.isEmpty()) {
            return DeliveryState(
                sectionVisible = false,
                options = emptyList(),
                headerText = "",
                minimumOrderTextPresent = false,
            )
        }
        return DeliveryState(
            sectionVisible = true,
            options = activeOptions.mapIndexed { i, opt -> DeliveryOption(opt.name, i == 0) },
            headerText = "Delivery Options",
            minimumOrderTextPresent = false,
        )
    }

    override fun getSearchResults(): SearchResults {
        val cards = fetchProducts(lastSearchTerm).map { it.toProductCard() }
        return SearchResults(cards, emptyStateVisible = cards.isEmpty())
    }

    override fun currentUrl(): String = when {
        lastSearchTerm.isNotBlank() -> "$BASE_URL/products?search=$lastSearchTerm"
        currentProductId >= 0      -> "$BASE_URL/products/$currentProductId"
        else                       -> "$BASE_URL/products"
    }

    private fun fetchProducts(searchTerm: String): List<ApiProduct> =
        if (searchTerm.isBlank())
            http.getListAs<ApiProduct>("/products")
        else
            http.getListWithQueryAs("/products", "search", searchTerm)

    private fun fetchProductDetail(id: Int): ApiProductDetail = http.getAs("/products/$id")

    private fun fetchActiveDeliveryOptions(): List<ApiDeliveryOption> =
        fetchProductDetail(currentProductId).deliveryOptions.filter { it.isActive }

    private fun ApiProduct.toProductCard() = ProductCard(
        title = title,
        price = formatPrice(price),
        imageUrl = imageUrl?.let { if (it.startsWith("http")) it else "$BASE_URL$it" } ?: "",
    )

    private fun formatPrice(price: Double) = "$${"%.2f".format(price)}"
}

File: typescript-automation/src/driver/api/myEcommerceDriver.ts
import type { HttpPort } from "../ports/httpPort";
import type { ApiDeliveryOption, ApiProduct, ApiProductDetail } from "./apiModels";
import type { CatalogueProtocol } from "../../dsl/protocols/catalogueProtocol";
import { register } from "../../dsl/protocols/driverRegistry";
import type { DeliveryOption, DeliveryState, ProductCard, ProductDetail, ProductListing, SearchResults } from "../../dsl/domain/index";
import { AxiosHttpPort } from "./axiosHttpPort";

const BASE_URL = "http://localhost:8001";

register("API", () => new MyEcommerceDriver(new AxiosHttpPort(BASE_URL)));

export class MyEcommerceDriver implements CatalogueProtocol {
  private currentProductId = -1;
  private lastSearchTerm = "";

  constructor(private readonly http: HttpPort) {}

  async browseCatalogue(): Promise<void> {}
  async returnToProductListing(): Promise<void> {}
  async chooseAlternativeDeliveryOption(): Promise<void> {}

  async viewFirstProduct(): Promise<void> {
    const products = await this.fetchProducts("");
    const first = products[0];
    if (first === undefined) throw new Error("No products available");
    this.currentProductId = first.id;
  }

  async searchFor(term: string): Promise<void> {
    this.lastSearchTerm = term;
  }

  async getProductListing(): Promise<ProductListing> {
    const cards = (await this.fetchProducts("")).map((p) => this.toProductCard(p));
    return { cards, hasVisibleLoadingIndicators: false };
  }

  async getProductDetail(): Promise<ProductDetail> {
    const product = await this.fetchProductDetail(this.currentProductId);
    return {
      title: product.title,
      price: this.formatPrice(product.price),
      description: product.description,
      imagePresent: product.image_url !== null && product.image_url !== "",
      addToCartButtonText: "Add to Cart",
      addToCartEnabled: true,
    };
  }

  async getDeliveryState(): Promise<DeliveryState> {
    const activeOptions = await this.fetchActiveDeliveryOptions();
    if (activeOptions.length === 0) {
      return { sectionVisible: false, options: [], headerText: "", minimumOrderTextPresent: false };
    }
    const options: ReadonlyArray<DeliveryOption> = activeOptions.map((opt, i) => ({
      label: opt.name,
      selected: i === 0,
    }));
    return { sectionVisible: true, options, headerText: "Delivery Options", minimumOrderTextPresent: false };
  }

  async getSearchResults(): Promise<SearchResults> {
    const cards = (await this.fetchProducts(this.lastSearchTerm)).map((p) => this.toProductCard(p));
    return { cards, emptyStateVisible: cards.length === 0 };
  }

  async currentUrl(): Promise<string> {
    if (this.lastSearchTerm !== "") return `${BASE_URL}/products?search=${this.lastSearchTerm}`;
    if (this.currentProductId >= 0) return `${BASE_URL}/products/${this.currentProductId}`;
    return `${BASE_URL}/products`;
  }

  private async fetchProducts(searchTerm: string): Promise<ReadonlyArray<ApiProduct>> {
    if (searchTerm === "") return this.http.getListAs<ApiProduct>("/products");
    return this.http.getListWithQueryAs<ApiProduct>("/products", "search", searchTerm);
  }

  private fetchProductDetail(id: number): Promise<ApiProductDetail> {
    return this.http.getAs<ApiProductDetail>(`/products/${id}`);
  }

  private async fetchActiveDeliveryOptions(): Promise<ReadonlyArray<ApiDeliveryOption>> {
    const detail = await this.fetchProductDetail(this.currentProductId);
    return detail.delivery_options.filter((o) => o.is_active);
  }

  private toProductCard(product: ApiProduct): ProductCard {
    const rawUrl = product.image_url;
    const imageUrl = rawUrl === null || rawUrl === ""
      ? ""
      : rawUrl.startsWith("http") ? rawUrl : `${BASE_URL}${rawUrl}`;
    return { title: product.title, price: this.formatPrice(product.price), imageUrl };
  }

  private formatPrice(price: number): string {
    return `$${price.toFixed(2)}`;
  }
}
